authz/rbac_translator.go

```diff
@@ -0,0 +1,301 @@
+		}
+	}
+	return &v3routepb.HeaderMatcher{
+		Name:                 key,
+		HeaderMatchSpecifier: &v3routepb.HeaderMatcher_ExactMatch{ExactMatch: value},
+	}
+}
+
+func parsePrincipalNames(principalNames []string) []*v3rbacpb.Principal {
+	var or []*v3rbacpb.Principal
+	for _, principalName := range principalNames {
+		newPrincipalName := &v3rbacpb.Principal{
+			Identifier: &v3rbacpb.Principal_Authenticated_{
+				Authenticated: &v3rbacpb.Principal_Authenticated{
+					PrincipalName: getStringMatcher(principalName),
+				},
+			}}
+		or = append(or, newPrincipalName)
+	}
+	return or
+}
+
+func parsePeer(source peer) *v3rbacpb.Principal {
+	var and []*v3rbacpb.Principal
+	if source.Principals != nil {
+		or := parsePrincipalNames(source.Principals)
+		if len(or) > 0 {
+			and = append(and, principalOr(or))
+		}
+	}
+	if len(and) > 0 {
+		return principalAnd(and)
+	}
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_Any{
+			Any: true,
+		},
+	}
+}
+
+func parsePaths(paths []string) []*v3rbacpb.Permission {
+	var or []*v3rbacpb.Permission
+	for _, path := range paths {
+		newPath := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_UrlPath{
+				UrlPath: &v3matcherpb.PathMatcher{
+					Rule: &v3matcherpb.PathMatcher_Path{Path: getStringMatcher(path)}}}}
+		or = append(or, newPath)
+	}
+	return or
+}
+
+func parseHeaderValues(key string, values []string) []*v3rbacpb.Permission {
+	var or []*v3rbacpb.Permission
+	for _, value := range values {
+		newHeader := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_Header{
+				Header: getHeaderMatcher(key, value)}}
+		or = append(or, newHeader)
+	}
+	return or
+}
+
+func parseHeaders(headers []header) ([]*v3rbacpb.Permission, error) {
+	var and []*v3rbacpb.Permission
+	for i, header := range headers {
+		if header.Key == "" {
+			return nil, fmt.Errorf("\"headers\" %d: \"key\" is not present", i)
+		}
+		// TODO(ashithasantosh): Return error for unsupported headers- "hop-by-hop",
+		// pseudo headers, "Host" header and headers prefixed with "grpc-".
+		if header.Values == nil {
+			return nil, fmt.Errorf("\"headers\" %d: \"values\" is not present", i)
+		}
+		and = append(and, permissionOr(parseHeaderValues(header.Key, header.Values)))
+	}
+	return and, nil
+}
+
+func parseRequest(request request) (*v3rbacpb.Permission, error) {
+	var and []*v3rbacpb.Permission
+	if request.Paths != nil {
+		or := parsePaths(request.Paths)
+		if len(or) > 0 {
+			and = append(and, permissionOr(or))
+		}
+	}
+	if request.Headers != nil {
+		subAnd, err := parseHeaders(request.Headers)
+		if err != nil {
+			return nil, err
+		}
+		if len(subAnd) > 0 {
+			and = append(and, permissionAnd(subAnd))
+		}
+	}
+	if len(and) > 0 {
+		return permissionAnd(and), nil
+	}
+	return &v3rbacpb.Permission{
+		Rule: &v3rbacpb.Permission_Any{
+			Any: true,
+		},
+	}, nil
+}
+
+func parseRulesArray(rules []rule, prefixName string) (map[string]*v3rbacpb.Policy, error) {
```

Comment:
`parseRules`?  Go has slices -- and technically arrays, too, but nobody uses them.